modules/openid_connect/app/models/openid_connect/provider/hash_builder.rb

```diff
@@ -30,55 +30,93 @@
 
 module OpenIDConnect
   module Provider::HashBuilder
-    def attribute_map
-      OpenIDConnect::Provider::MAPPABLE_ATTRIBUTES
-        .index_with { |attr| public_send(:"mapping_#{attr}") }
-        .compact_blank
+    BACKCHANNEL_LOGOUT_CALLBACK = ->(logout_token) do
+      ::OpenProject::OpenIDConnect::SessionMapper.handle_logout(logout_token)
     end
 
-    def to_h # rubocop:disable Metrics/AbcSize
+    STATE_GENERATOR = -> { SecureRandom.hex(42) }
+
+    def to_omniauth_hash # rubocop:disable Metrics/AbcSize
+      client_options = {
+        identifier: client_id,
+        secret: client_secret,
+        authorization_endpoint:,
+        token_endpoint:,
+        userinfo_endpoint:,
+        jwks_uri:,
+        end_session_endpoint:,
+        host: host || ::Storages::UrlBuilder.host_from_endpoint(authorization_endpoint),
+        redirect_uri:
+      }.compact
       {
-        name: slug,
-        oidc_provider:,
+        name: slug.to_sym,
         icon:,
-        host:,
         scheme:,
         port:,
         display_name:,
-        userinfo_endpoint:,
-        authorization_endpoint:,
-        jwks_uri:,
         issuer:,
-        scope:,
-        identifier: client_id,
-        secret: client_secret,
-        token_endpoint:,
-        limit_self_registration:,
-        end_session_endpoint:,
+        scope: scopes,
         attribute_map:,
         post_logout_redirect_uri:,
-        claims:,
-        acr_values:
+        claims: add_groups_claim(JSON.parse(claims.presence || "{}")).to_json,
+        acr_values:,
+        client_options:
       }
-       .merge(provider_specific_to_h)
-       .compact_blank
+        .merge(provider_specific_to_h)
+        .compact
+    end
+
+    def to_h
+      hash = to_omniauth_hash
+      hash.merge(
+        limit_self_registration:,
+        # Remember oidc session values when logging in user
+        retain_from_session: %w[
+          omniauth.oidc_sid
+          omniauth.oidc_access_token
+          omniauth.oidc_refresh_token
+          omniauth.oidc_expires_in
+          omniauth.oidc_groups
+        ],
+        single_sign_out_callback: single_sign_out_callback(hash),
+        backchannel_logout_callback: BACKCHANNEL_LOGOUT_CALLBACK
```

Comment:
We might be able to completely get rid of this constant. Since the callback is only implemented as "call this method", we should be able to write this as:

```suggestion
        backchannel_logout_callback: ::OpenProject::OpenIDConnect::SessionMapper.method(:handle_logout)
```